core/src/main/scala/org/typelevel/ci/CaseFoldedString.scala

```diff
@@ -0,0 +1,162 @@
+  * transformation. You should use [[CaseFoldedString]] instead of
+  * [[CIString]] when you only want the case insensitive `String` and you
+  * never want to return the `String` back into the input value. In such cases
+  * [[CaseFoldedString]] will be more efficient than [[CIString]] as it only
+  * has to keep around a single `String` in memory.
+  *
+  * Case insensitive `String` values under Unicode are not always intuitive,
+  * especially on the JVM. There are three character cases to consider, lower
+  * case, upper case, and title case, and not all Unicode codePoints have all
+  * 3, some only have 2, some only 1. For some codePoints, the JRE standard
+  * operations don't always work as you'd expect.
+  *
+  * {{{
+  * scala> val codePoint: Int = 8093
+  * val codePoint: Int = 8093
+  *
+  * scala> new String(Character.toChars(codePoint))
+  * val res0: String = ᾝ
+  *
+  * scala> res0.toUpperCase
+  * val res1: String = ἭΙ
+  *
+  * scala> res0.toUpperCase.toLowerCase == res0.toLowerCase
+  * val res2: Boolean = false
+  *
+  * scala> Character.getName(res0.head)
+  * val res3: String = GREEK CAPITAL LETTER ETA WITH DASIA AND OXIA AND PROSGEGRAMMENI
+  *
+  * scala> res0.toUpperCase.toLowerCase.equalsIgnoreCase(res0.toLowerCase)
+  * val res4: Boolean = false
+  * }}}
+  *
+  * In this example, given the Unicode character \u1f9d, converting it to
+  * upper case, then to lower case, is not equal under normal String
+  * equality. `String.equalsIgnoreCase` also does not work correctly by the
+  * Unicode standard.
+  *
+  * Making matters more complicated, for certain Turkic languages, the case
+  * folding rules change. See the Unicode standard for a full discussion of
+  * the topic.
+  *
+  * @note For most `String` values the `toString` form of this is lower case
+  *       (when the given character has more than one case), but this is not
+  *       always the case. Certain Unicode scripts have exceptions to this and
+  *       will be case folded into upper case. If you want/need an only lower
+  *       case `String`, you should call `.toString.toLowerCase`.
+  *
+  * @see [[https://www.unicode.org/versions/Unicode14.0.0/ch05.pdf#G21790]]
+  */
+final case class CaseFoldedString private (override val toString: String) extends AnyVal {
+
+  def isEmpty: Boolean = toString.isEmpty
+
+  def nonEmpty: Boolean = !isEmpty
+
+  def length: Int = toString.length
+
+  def size: Int = length
+
+  def trim: CaseFoldedString =
+    CaseFoldedString(toString.trim)
+
+  private final def copy(toString: String): CaseFoldedString =
+    CaseFoldedString(toString)
+}
+
+object CaseFoldedString {
+
+  /** Create a [[CaseFoldedString]] from a `String`.
+    *
+    * @param turkicFoldingRules if `true`, use the case folding rules for
+    *                           applicable to some Turkic languages.
+    */
+  def apply(value: String, turkicFoldingRules: Boolean): CaseFoldedString = {
```

Comment:
How about just calling this one `turkic` and losing the boolean (and the overload).  Are people going to need to toggle this with a boolean?

Reply:
I know I've created a bit of a mess by creating this and #232, but in that one I create separate types for Turkic and non-Turkic variants. I think _if_ we support Turkic variants, we need to have separate types. I'm concerned that allowing two `CaseFoldedString` values to have the same `String` representation, but be created under different rules will be a source of rare, but terrible, bugs.

Reply:
I think they would need to be separate types.  I guess we can support them, but it's locale-like behavior, and I'm not sure where it ends.

Reply:
Well...if we want to support Simple/Full Default/Canonical/Compatibility caseless matching, I feel like the Turkic variants aren't really the primary issue. 

Unfortunately, it sounds like there are quite real use cases for most of these permutations :slightly_frowning_face: